routers/api/actions/artifacts.go

```diff
@@ -0,0 +1,293 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package actions
+
+import (
+	gocontext "context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"code.gitea.io/gitea/models/actions"
+	"code.gitea.io/gitea/modules/context"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/storage"
+	"code.gitea.io/gitea/modules/web"
+)
+
+const artifactRouteBase = "/_apis/pipelines/workflows/{taskID}/artifacts"
+
+func ArtifactsRoutes(goctx gocontext.Context, prefix string) *web.Route {
+	m := web.NewRoute()
+	m.Use(withContexter(goctx))
+
+	r := artifactRoutes{
+		prefix: prefix,
+		fs:     storage.ActionsArtifacts}
+
+	// retrieve, list and confirm artifacts
+	m.Post(artifactRouteBase, r.getUploadArtifactURL)
+	m.Get(artifactRouteBase, r.listArtifacts)
+	m.Patch(artifactRouteBase, r.patchArtifact)
+
+	// handle container artifacts
+	m.Put(artifactRouteBase+"/{artifactID}/upload", r.uploadArtifact)
+	m.Get(artifactRouteBase+"/{artifactID}/path", r.getDownloadArtifactURL)
+	m.Get(artifactRouteBase+"/{artifactID}/download", r.downloadArtifact)
+
+	return m
+}
+
+// withContexter initializes a package context for a request.
+func withContexter(ctx gocontext.Context) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(resp http.ResponseWriter, req *http.Request) {
+			ctx := context.Context{
+				Resp: context.NewResponse(resp),
+				Data: map[string]interface{}{},
+			}
+			defer ctx.Close()
+
+			ctx.Req = context.WithContext(req, &ctx)
+
+			next.ServeHTTP(ctx.Resp, ctx.Req)
+		})
+	}
+}
+
+type artifactRoutes struct {
+	prefix string
+	fs     storage.ObjectStorage
+}
+
+func (ar artifactRoutes) openFile(fpath string, contentRange string) (storage.Object, bool, error) {
+	if contentRange != "" && !strings.HasPrefix(contentRange, "bytes 0-") {
+		f, err := ar.fs.Open(fpath)
+		if err != nil {
+			return nil, false, err
+		}
+		_, err = f.Seek(0, os.SEEK_END)
+		return f, true, err
+	}
+	f, err := ar.fs.Open(fpath)
+	return f, false, err
+}
+
+func (ar artifactRoutes) buildArtifactUrl(taskID int64, artifactID int64, suffix string) (string, error) {
+	uploadURL := strings.TrimSuffix(setting.AppURL, "/") + strings.TrimSuffix(ar.prefix, "/") +
+		strings.ReplaceAll(artifactRouteBase, "{taskID}", strconv.FormatInt(taskID, 10)) +
+		"/" + strconv.FormatInt(artifactID, 10) + "/" + suffix
+	u, err := url.Parse(uploadURL)
+	if err != nil {
+		return "", err
+	}
+	// FIXME: fix localhost ?? act starts docker container to run action.
+	// It can't visit host network in container.
+	if strings.Contains(u.Host, "localhost") {
+		u.Host = strings.ReplaceAll(u.Host, "localhost", actions.GetOutboundIP().String())
```

Comment:
It's too expensive to have a DNS query.

Reply:
Even worse, if the guess is wrong (say, inside the docker), there will more problems than it resolves, it confuses users and developers if they see the strange error message with a strange IP.

Reply:
I think we can just ignore this.